### PR TITLE
Add a new COMContext object in AppInstaller Core to enable COM Interface to consume AppInstaller

### PR DESCRIPTION
**Pull request by sachintaMSFT**
Add a new COMContext object in AppInstaller Core to enable COM Interface to issue any commands to AppInstaller. The COM Context object will also enable COM Interface to see the progress, receive errors and cancel execution of commands, where applicable.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/818)

### CODE REVIEW

**Comment by sachintaMSFT** [comment]
> #### Misspellings found, please review:
> * streambuf
> 
> To accept these changes, run the following commands from this repository on this branch

will add streambuf word to .github/actions/spelling/expect.txt in alphabetical sort order along in the commit for incorporation of any review comments received.

File: src/AppInstallerCLICore/Public/AppInstallerCLICoreExecute.h [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+#include "ExecutionContext.h"
+#include "Command.h"
+
+namespace AppInstaller::CLI
+{
+    int Execute(Execution::Context& context, std::unique_ptr<Command>& command);
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Public/AppInstallerCLICoreExecute.h:7]: Assuming that the plan is to implement the COM interface classes in another static lib (I support that plan), then many other headers will "need" to be placed in public.  I think it makes sense for Command and Context to be made public, and then COMContext will actually be moved over to that lib.

This function should then be in Command.h, and the "make this command" function should be too.

For now, COMContext can live where you have it in this PR, but at least move this function to Command.h.

File: src/AppInstallerCLICore/Core.cpp [hunk 2]
```diff
@@ -115,6 +117,17 @@ namespace AppInstaller::CLI
...
+        return Execute(context, command);
+    }
+    // End of the line exceptions that are not ever expected.
+    // Telemetry cannot be reliable beyond this point, so don't let these happen.
+    catch (...)
+    {
+        return APPINSTALLER_CLI_ERROR_INTERNAL_ERROR;
+    }
+
+    int Execute(Execution::Context& context, std::unique_ptr<Command>& command)
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Core.cpp:129]: RE: my other comment, move this to Command.cpp.

File: src/AppInstallerCLICore/ExecutionContext.h [hunk 3]
```diff
@@ -101,6 +101,10 @@ namespace AppInstaller::CLI::Execution
             WI_ClearAllFlags(m_flags, flags);
         }
 
+        void SetExecutionStage(Workflow::ExecutionStage stage, bool allowBackward);
+        virtual void SetExecutionStage(Workflow::ExecutionStage stage);
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/ExecutionContext.h:105]: No need to separate these functions, but if you are going to do so, the virtual function should be protected.

File: src/AppInstallerCLITests/TestCommon.cpp [hunk 4]
```diff
@@ -139,6 +139,15 @@ namespace TestCommon
         }
     }
 
+    void TestProgress::BeginProgress()
+    {
+    }
+
+    void TestProgress::EndProgress(bool hideProgressWhenDone)
+    {
+        _Unreferenced_parameter_(hideProgressWhenDone);
```
Comment (JohnMcPMS) [inline src/AppInstallerCLITests/TestCommon.cpp:148]: Remove name from signature.

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]

File: src/AppInstallerCLICore/COMContext.cpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    NullStream::NullStream()
+    {
+        null_Out.reset(new std::ostream(&nullStreamBuf));
+        null_In.reset(new std::istream(&nullStreamBuf));
+    }
+
+    void COMContext::BeginProgress()
+    {
+        m_comProgressCallback(ReportType::BeginProgress, m_current, m_maximum, m_type, m_executionPhase);
```
[the remark this answers is not in the document]
Comment (sachintaMSFT) [reply src/AppInstallerCLICore/COMContext.cpp:17]: I do reset progress values in SetExecutionStage. But I guess if there are multiple separate & sequential progresses in same ExecutionStage (which I think Discover stage has), that doesn't help BeginProgress send reset progress values. And if that progress is super quick that OnProgress isn't called and EndProgress get called directly, then the leak goes into EndProgress as well. Accept your suggestion.

**Review by JohnMcPMS — APPROVED** [review]

**Pull request merged**